presto-main/src/main/java/com/facebook/presto/cost/ScalarStatsCalculator.java

```diff
@@ -228,5 +229,44 @@ private double operate(ArithmeticBinaryExpression.Type type, double left, double
                     throw new IllegalStateException("Unsupported ArithmeticBinaryExpression.Type: " + type);
             }
         }
+
+        @Override
+        protected SymbolStatsEstimate visitCoalesceExpression(CoalesceExpression node, Void context)
+        {
+            requireNonNull(node, "node is null");
+            SymbolStatsEstimate result = null;
+            for (Expression operand : node.getOperands()) {
+                SymbolStatsEstimate operandEstimates = process(operand);
+                if (result != null) {
+                    result = estimateCoalesce(result, operandEstimates);
+                }
+                else {
+                    result = operandEstimates;
+                }
+            }
+            return requireNonNull(result, "result is null");
+        }
+
+        private SymbolStatsEstimate estimateCoalesce(SymbolStatsEstimate left, SymbolStatsEstimate right)
+        {
+            // Question to reviewer: do you have a method to check if fraction is empty or saturated?
+            if (left.getNullsFraction() == 0) {
+                return left;
+            }
+            else if (left.getNullsFraction() == 1.0) {
+                return right;
+            }
+            else {
+                return SymbolStatsEstimate.builder()
+                        .setLowValue(min(left.getLowValue(), right.getLowValue()))
+                        .setHighValue(max(left.getHighValue(), right.getLowValue()))
+                        .setDistinctValuesCount(left.getDistinctValuesCount() +
```

Comment:
So this assumes that left and right distinct values are all distinct from each other too?  Seems like it should depend on how much their ranges overlap and the #rows/range size ratio.

Reply:
I understand this idea in a way resembling how we handle `=` stats (in Join or Filter). There, we assume all NDVs within overlapping ranges to be the same. Of course, this is just a heuristic. What heuristic would be most applicable for `coalesce`?

Reply:
yeah, you're right.  I was inspired by equals, but there's no reason the heuristics should be the same.  My instinct (based neither on real data nor anecdotal evidence) says that double types are less likely to be the same because the number of possible values in a given range is huge, and since the user doesn't say to compare them there's no reason to assume they match up exactly.  Integer types seem more likely to be similar to equals.   Anyway, probably better to overestimate than underestimate, so this is fine.